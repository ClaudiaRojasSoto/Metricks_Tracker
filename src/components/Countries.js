import React, { useEffect } from 'react';
import { useParams, Link } from 'react-router-dom';
import { useSelector, useDispatch } from 'react-redux';
import axios from 'axios';
import { updateCountries, updateCountryData } from '../redux/actions/countriesActions';
import flagsData from '../flagData';
import '../styles/Countries.css';

import continent1 from '../imagesContinents/continents/africa.png';
import continent2 from '../imagesContinents/continents/america.png';
import continent3 from '../imagesContinents/continents/asia.png';
import continent4 from '../imagesContinents/continents/australia.png';
import continent5 from '../imagesContinents/continents/europa.png';
import continent6 from '../imagesContinents/continents/oceania.png';

function getContinentImage(continent) {
  switch (continent) {
    case 'africa':
      return continent1;
    case 'america':
      return continent2;
    case 'asia':
      return continent3;
    case 'australia':
      return continent4;
    case 'europa':
      return continent5;
    case 'oceania':
      return continent6;
    default:
      return null;
  }
}

const Countries = () => {
  const { continent } = useParams();
  const countries = useSelector((state) => state.countries.countries);
  const dispatch = useDispatch();

  useEffect(() => {
    const countriesOfContinent = flagsData[continent];
    dispatch(updateCountries(countriesOfContinent));
  }, [continent, dispatch]);

  if (!Array.isArray(countries)) {
    return (
      <div>
        <h1>{`Error: ${continent} It is not a valid continent.`}</h1>
      </div>
    );
  }

  const handleCountryClick = async (country) => {
    try {
      const response = await axios.get(`https://restcountries.com/v3/alpha/${country.countryCode}`);
      if (response.data && response.data.length > 0) {
        console.log(response.data);
        dispatch(updateCountryData(response.data[0]));
      } else {
        console.log('No se devolvieron datos válidos del país en la API');
      }
    } catch (error) {
      console.log('Error al obtener los datos del país:', error);
    }
  };

  const capitalizeFirstLetter = (
    string,
  ) => string.charAt(0).toUpperCase() + string.slice(1).toLowerCase();

  return (
    <div>
      <img src={getContinentImage(continent)} alt={continent} />
      <h1>{capitalizeFirstLetter(continent)}</h1>
      <div className="countries-grid">
        {countries.map((country) => (
          <Link
            to={{
              pathname: `/country/${country.name.common}`,
              state: { countryData: country },
              // Pasamos toda la información del país en el estado de la ruta
            }}
            key={country.countryCode}
          >
            <div
              className="country-item"
              onClick={() => handleCountryClick(country)}
              onKeyDown={(e) => {
                if (e.key === 'Enter' || e.key === ' ') handleCountryClick(country);
              }}
              tabIndex="0"
              role="button"
            >
              <img src={country.imagePath} alt={country.name} />
              <p>{country.name}</p>
            </div>
          </Link>
        ))}
      </div>
    </div>
  );
};

export default Countries;
