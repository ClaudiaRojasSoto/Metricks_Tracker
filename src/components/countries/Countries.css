.countries-grid {
  display: grid;
  grid-template-columns: repeat(2, 1fr); /* Dos columnas con tamaño igual */
  gap: 20px; /* Espacio entre los elementos */
}

.country-item {
  display: flex;
  flex-direction: column;
  align-items: center;
  text-align: center;
}

img {
  width: 100px; /* Ajusta el tamaño de la bandera */
  height: auto;
}
