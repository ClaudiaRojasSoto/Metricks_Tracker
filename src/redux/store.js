import { configureStore } from '@reduxjs/toolkit';
import countriesReducer from './reducers/countriesReducer';

const store = configureStore({
  reducer: {
    countries: countriesReducer,
  },
});

export default store;
