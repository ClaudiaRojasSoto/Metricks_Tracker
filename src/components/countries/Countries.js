import React, { useEffect, useState } from 'react';
import { useParams, Link } from 'react-router-dom';
import { useSelector, useDispatch } from 'react-redux';
import { fetchCountriesByContinent, updateCountryData } from '../../redux/actions/countriesActions';
import './Countries.css';

const Countries = () => {
  const { continent } = useParams();
  const countries = useSelector((state) => state.countries.countries);
  const dispatch = useDispatch();

  const [filter, setFilter] = useState('');

  useEffect(() => {
    dispatch(fetchCountriesByContinent(continent));
  }, [continent, dispatch]);

  const handleCountryClick = async (country) => {
    dispatch(updateCountryData(country));
  };

  const handleFilterChange = (event) => {
    setFilter(event.target.value);
  };

  const filteredCountries = countries.filter((
    country,
  ) => country.name.common.toLowerCase().includes(filter.toLowerCase()));

  const capitalizeFirstLetter = (string = '') => string.charAt(0).toUpperCase() + string.slice(1).toLowerCase();

  return (
    <div>
      <h1>{capitalizeFirstLetter(continent)}</h1>
      <input
        type="text"
        value={filter}
        onChange={handleFilterChange}
        placeholder="Filtrar por nombre de paÃ­s..."
      />
      <div className="countries-grid">
        {filteredCountries.map((country) => (
          <Link
            to={{
              pathname: `/country/${country.name.common}`,
              state: { countryData: country },
            }}
            key={country.cca3}
          >
            <div
              className="country-item"
              onClick={() => handleCountryClick(country)}
              onKeyDown={(e) => {
                if (e.key === 'Enter' || e.key === ' ') handleCountryClick(country);
              }}
              tabIndex="0"
              role="button"
            >
              <img src={country.flags.png} alt={country.name.common} />
              <p>{country.name.common}</p>
            </div>
          </Link>
        ))}
      </div>
    </div>
  );
};

export default Countries;
