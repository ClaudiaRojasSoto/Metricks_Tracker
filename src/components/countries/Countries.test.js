import '@testing-library/jest-dom';
import { BrowserRouter as Router } from 'react-router-dom';
import renderer from 'react-test-renderer';
import { Provider } from 'react-redux';
import store from '../../redux/store';
import Countries from './Countries';

describe('Countries tests', () => {
  let countries;
  beforeAll(async () => {
    countries = renderer.create(
      <Provider store={store}>
        <Router>
          <Countries />
        </Router>
      </Provider>,
    ).toJSON();
  });

  it('Component should render correctly', () => {
    expect(countries).toBeTruthy();
  });

  it('It should match its snapshot', () => {
    expect(countries).toMatchSnapshot();
  });
});
