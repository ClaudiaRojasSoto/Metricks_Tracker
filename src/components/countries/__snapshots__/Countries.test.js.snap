// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Countries tests It should match its snapshot 1`] = `
<div>
  <div
    className="search"
  >
    <h1 />
    <input
      onChange={[Function]}
      placeholder="Filter by country name..."
      type="text"
      value=""
    />
  </div>
  <div
    className="countries_container"
  />
</div>
`;
